text/3127-trim-paths.md

```diff
@@ -0,0 +1,255 @@
+
+- `macro` - apply remappings to the expansion of `std::file!()` macro. This is where paths in embedded panic messages come from
+- `debuginfo` - apply remappings to debug information, wherever they may be written to
+- `split-debuginfo-file` - when `split-debuginfo=packed` or `unpacked`, apply remappings to the paths pointing to these split debug information files
+- `diagnostics` - apply remappings to printed compiler diagnostics
+
+## Cargo
+
+`trim-paths` is a profile setting which controls the sanitisation of file paths in compilation outputs. It has three valid options:
+- `0` or `false`: no sanitisation at all
+- `1`: sanitise only the paths in emitted executable or library binaries. It always affects paths from macros such as panic messages, and in debug information
+  only if they will be embedded together with the binary (the default on platforms with ELF binaries, such as Linux and windows-gnu),
+  but will not touch them if they are in separate files (the default on Windows MSVC and macOS). But the path to these separate files are sanitised.
+- `2` or `true`: sanitise paths in all compilation outputs, including compiled executable/library, debug information, and compiler diagnostics.
+
+The default release profile uses option `1`. You can also manually override it by specifying this option in `Cargo.toml`:
+```toml
+[profile.dev]
+trim-paths = 2
+
+[profile.release]
+trim-paths = 0
+```
+
+When a path is in scope for sanitisation, it is handled by the following rules:
+
+1. Path to the source files of the standard and core library (sysroot) will begin with `/rustc/[rustc commit hash]`.
+   E.g. `/home/username/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs` -> 
+   `/rustc/fe72845f7bb6a77b9e671e6a4f32fe714962cec4/library/core/src/result.rs`
+2. Path to the working directory will be stripped. E.g. `/home/username/crate/src/lib.rs` -> `src/lib.rs`.
+3. Path to packages outside of the working directory will be replaced with `[package name]-[version]`. E.g. `/home/username/deps/foo/src/lib.rs` -> `foo-0.1.0/src/lib.rs`
+
+When a path to the source files of the standard and core library is *not* in scope for sanitisation, the emitted path will depend on if `rust-src` component
+is present. If it is, then the real path pointing to a copy of the source files on your file system will be emitted; if it isn't, then they will
+show up as `/rustc/[rustc commit hash]/library/...` (just like when it is selected for sanitisation). Paths to all other source files will not be affected.
+
+This will not affect any hard-coded paths in the source code.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `trim-paths` implementation in Cargo
+
+We only need to change the behaviour for `Test` and `Build` compile modes. 
+
+If `trim-paths` is `0` (`false`), no extra flag is supplied to `rustc`.
```

Comment:
I would propose that, rather than using numbers here (which are not forwards-compatible if we want to add more levels in between), we should use descriptive terms:

- `trim-paths=none` or `trim-paths=false` (don't trim any paths)
- `trim-paths=object` (trim paths only within the compiled object)
- `trim-paths=all` or `trim-paths=true` (trim all paths, including diagnostics, split debuginfo files, etc)

This would allow us to, in the future, add more names for useful subsets.

I'm also wondering if this could then be merged into `remap-path-scopes`; we could just add a `remap-path-scopes=object` as an alias for "everything included in the object", and a `remap-path-scopes=all` to include everything. That would make it convenient to write things like `remap-path-scopes=object,debuginfo` if you want to remap paths in the object and in split debuginfo but not in diagnostics.

Cargo could then just offer a single option that takes a list of those scopes or aliases for combinations of scopes, and passes them through to rustc.

Reply:
I agree with using strings instead of numbers.

I've added the alias idea to future possibilities. It looks convenient but I guess we could wait and see if people would like to use it. This can be implemented at anytime without backwards compatibility issues

Reply:
@cbeuw The main reason to add it now would be to simplify Cargo's job, so that it can just pass through a value or list of values to rustc.

I'd personally prefer to have the `all` and `object` aliases as part of the initial RFC.

Reply:
I've added `all`, `true` and `object` aliases, so the option in Cargo's `trim-paths` can be supplied directly to `--remap-path-scope`, along with the two `--remap-path-prefix` mappings. Except for when `trim-paths` is `none` or `false`, then none of `--remap-path-scope` or `--remap-path-prefix` are supplied to rustc.

Should we also allow `--remap-path-scope` to take on `none` and `false`? In which case it disables all effects of `--remap-path-prefix`, though in this case one should always simply not supply anything.